### ISSUE TITLE
Support ChaCha20 ciphers.

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#8760]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)| @Lukasa reported|
|-|-|
|Trac ID|trac#8760|
|Type|enhancement|
|Created|2016-08-15 13:22:01Z|
|Branch|https://github.com/twisted/twisted/tree/8760-lukasa-chacha20|

ChaCha20 is coming in OpenSSL 1.1.0, and it's generally pretty excellent. OpenSSL kindly ignores requests for ciphers that it doesn't support, so we can unconditionally update our code to request ChaCha20 and expect OpenSSL to just quietly ignore the change when it doesn't know what ChaCha is.

This change has us prefer AES-GCM over ChaCha20. This decision is made mostly because newer processors have AES-NI hardware acceleration instructions, and right now we're not in a good place to query whether or not those instructions are present, so we just assume that they are. However, CPython issue 27766 (https://bugs.python.org/issue27766), in addition to making this change, also provides a private function that could be queried for AES-NI support. We may want to wait until CPython 3.6 drops to see if that function makes it in, because we could use it to be smarter about our cipher string.

However, I think we should merge this anyway. It's a strict improvement over what we already have.

<details><summary>Searchable metadata</summary>

```
trac-id__8760 8760
type__enhancement enhancement
reporter__Lukasa Lukasa
priority__normal normal
milestone__None None
branch__8760_lukasa_chacha20 8760-lukasa-chacha20
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1471267321624906 1471267321624906
changetime__1471426760730333 1471426760730333
version__None None
owner__Lukasa Lukasa

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_1"></a>|@Lukasa commented|
|-|-|

In changeset 6a214b8b18f0875c8816a4019483ee6e534ab226
```
#!CommitTicketReference repository="" revision="6a214b8b18f0875c8816a4019483ee6e534ab226"
Add topfile for #8760
```

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_2"></a>|@Lukasa commented|
|-|-|

PR for this is available here: https://github.com/twisted/twisted/pull/499

**Comment by twisted-trac** [comment]
|[<img alt="adiroiban's avatar" src="https://avatars.githubusercontent.com/u/204609?s=50" width="50" height="50">](https://github.com/adiroiban)<a name="note_3"></a>|@adiroiban set owner to @Lukasa|
|-|-|

Changes looks good.
Only a minor comment regarding the release notes.

Please check my commend and merge at will... as long as the test pass :)

Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="Lukasa's avatar" src="https://avatars.githubusercontent.com/u/1382556?s=50" width="50" height="50">](https://github.com/Lukasa)<a name="note_4"></a>|@Lukasa set status to `closed`|
|-|-|

In changeset 833abfdb340076ff79628d3c7631a92406899525
```
#!CommitTicketReference repository="" revision="833abfdb340076ff79628d3c7631a92406899525"
Merge 8760-lukasa-chacha20: Add support for ChaCha20.

Author: Lukasa
Reviewer: adiroiban
Fixes: #8760
```
